File: A2Z_striver/1 Basics/1.2 Patterns/pattern_11/pattern_11.py
import sys

class Solution:
    """
    Problem: Print a Binary Triangle (Alternating 1s and 0s).
    Example Input: n = 3
    Example Output:
    1
    0 1
    1 0 1
    Goal: O(N^2) Time.
    """

    # // ~ Approach 1: Coordinate Parity (Your Approach)
    # // Analysis: Time O(N^2) | Space O(1)
    # // Pros: Stateless, purely mathematical.
    def pattern11_parity(self, n):
        for i in range(n):
            for j in range(i + 1):
                # // ! Optimization: Move logic inside print
                # // Logic: Even sum -> 1, Odd sum -> 0
                print("1 " if (i + j) % 2 == 0 else "0 ", end="")
            print()

    # // ~ Approach 2: Toggle Method (State-based)
    # // Analysis: Time O(N^2) | Space O(1)
    # // Pros: Intuitive simulation of "flipping" a switch.
    def pattern11_toggle(self, n):
        for i in range(n):
            # // - Start with 1 if row is even, else 0
            start = 1 if i % 2 == 0 else 0
            
            for j in range(i + 1):
                print(f"{start} ", end="")
                # // - Flip the bit: 1->0, 0->1
                start = 1 - start 
            print()

# Driver Code
if __name__ == "__main__":
    solver = Solution()
    n = 3
    
    print(f"--- Approach 1: Parity (n={n}) ---")
    solver.pattern11_parity(n)
    
    print(f"\n--- Approach 2: Toggle (n={n}) ---")
    solver.pattern11_toggle(n)

File: A2Z_striver/1 Basics/1.2 Patterns/pattern_11/pattern_11.cpp
#include <iostream>

using namespace std;

// Header Comment Block
// Problem: Print a Binary Triangle (Alternating 1s and 0s).
// Example Input: n = 3
// Example Output:
// 1
// 0 1
// 1 0 1
// Goal: O(N^2) Time.

class Solution
{
public:
    // ~ Approach 1: The Toggle Method (Your approach, optimized)
    // Analysis: Time O(N^2) | Space O(1)
    // Pros: Easy to trace step-by-step.
    void pattern11_toggle(int n)
    {
        for (int i = 0; i < n; i++)
        {

            // - Step 1: Initialize start bit based on row parity
            //   Even row (0, 2) starts with 1. Odd row (1, 3) starts with 0.
            int start = (i % 2 == 0) ? 1 : 0;

            for (int j = 0; j <= i; j++)
            {
                cout << start << " ";

                // - Step 2: Flip the bit
                //   Trick: 1 - 1 = 0; 1 - 0 = 1;
                start = 1 - start;
            }
            cout << "\n";
        }
    }

    // ~ Approach 2: The Coordinate/Parity Method (Stateless)
    // Analysis: Time O(N^2) | Space O(1)
    // Pros: No state variable management; purely mathematical.
    void pattern11_parity(int n)
    {
        for (int i = 0; i < n; i++)
        {
            for (int j = 0; j <= i; j++)
            {

                // ! Logic: Check the sum of coordinates (i + j)
                // - If sum is EVEN -> Print 1
                // - If sum is ODD  -> Print 0
                if ((i + j) % 2 == 0)
                {
                    cout << "1 ";
                }
                else
                {
                    cout << "0 ";
                }
            }
            cout << "\n";
        }
    }
};

int main()
{
    ios_base::sync_with_stdio(false);
    cin.tie(NULL);

    Solution sol;
    int n = 5;

    cout << "--- Approach 1: Toggle Method ---\n";
    sol.pattern11_toggle(n);

    cout << "\n--- Approach 2: Parity Method ---\n";
    sol.pattern11_parity(n);

    return 0;
}